### ISSUE TITLE
Bug: Parser: End range of `Program` differs from Espree

### ISSUE THREAD

**Issue by overlookmotel** [typescript-eslint/typescript-eslint#11026]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&showAST=es&fileType=.tsx&code=PTAEDMHtIKBgPOJQCMCGAnGQ&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// foo

x

// bar
```

### ESLint Config

```javascript

```

### tsconfig

```jsonc

```

### Expected Result

The end location of `Program` differs between `@typescript-eslint/parser` and Espree.

Both TS-ESLint and Espree set the start location of `Program` as the first token, excluding whitespace and comments.

Espree sets the end location as the last token, also excluding whitespace and comments. However, `@typescript-eslint/parser` does not match that behavior - end location of `Program` is the end of the file.

I assume that `@typescript-eslint/parser` is aiming to align with Espree, and therefore this difference is unintentional.

Acorn does not exclude whitespace/comments from *either* start or end location. It's unfortunate the the 2 most-used ESTree-compatible parsers do not match each others' behavior, and personally I prefer the Acorn way - I think it's surprising that you can have comments which have locations *outside* the `Program` which contains them. But I imagine quite possible that neither Acorn nor Espree may want to make a breaking change.

Examples:

* [@typescript-eslint/parser](https://ast-explorer.dev/#eNolyTEKgDAQRNGrhKkVew9itc0aVojETUiCKOLd3WAzMO8/iJix88nVl5AbBmSDdmf5YZQag3b35tPktpRISa8+dlcupFaT1YfUOYL4gxcpNSQlzAaRm9RGIH3xfmfKJYM=)
* [Espree](https://ast-explorer.dev/#eNo9jDEKwzAMRa9iNAey5xylkxbVVUuKIxnZLi0hd6/chCxC7z34KySY4EVvKtHmXGGA7IJLNmaH6DCO4aGKgvLpx/FGhuJVva4oISAUbRb58s2MMDkvem/J/2HPHBe6spVZZe+JKpd69qSx+2qND2Mkz/9WdygbbD9jZzcY)
* [Acorn](https://ast-explorer.dev/#eNoliUEKhEAMBL8y5Cx49x3LnnLJZiMoYzJkRlHEvxvx0lRXnZBhgJk2quxTadBBCUFsrsEc3PdpNENF3Z+J+yPHp1rUEzUlBOGFvuJ1MkUYQmRqUhtC9/Zqq7N8jiJvXuy/5mDUC64b7zcqxA==)

### Actual Result

See above.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `8.29.0` |

**Comment by bradzacher** [comment]
It sounds to me like espree is wrong here?
Why would the range not include the comments and whitespace?

**Comment by overlookmotel** [comment]
> Why would the range not include the comments and whitespace?

That is my feeling too. I'll raise an issue on Espree and see what they say.

However, at present `@typescript-eslint/parser`'s behavior is half-and-half. It excludes whitespace/comments from the start range, but includes them in the end range. Whichever (include or exclude) is the "right" way to do it, the current behavior seems inconsistent.

**Comment by JLHwung** [comment]
Here is another case: with `ranges` option enabled:

```js
// single comment
```

In espree: `program.ranges` are `[0, 17]` while ts-eslint gives `[17, 17]`

**Comment by overlookmotel** [comment]
> Here is another case: with ranges option enabled:
> ```js
> // single comment
> ```

This is a great example! The fact that it's not clear what the "right" answer is suggests there's a problem with the mental model. I'd have expected `[0, 0]`.

Note: Esprima (which Espree aims to follow) is also `[17, 17]`!

Have opened an issue in Espree's repo: https://github.com/eslint/js/issues/648

**Comment by bradzacher** [comment]
IMO comments are part of the `Program` -- whether there is code or not.
So I'd personally say that us reporting `[17, 17]` is wrong.

What you guys are probably finding as you work on alignment is that when you look at ranges and tokens things get very messy.

The main problem with the ESTree spec is that it solely specifies AST shape and nothing more. So ranges, locs, tokens are all left as an exercise for the parser to guess at.

**Comment by fisker** [comment]
I think it's already discussed in espree https://github.com/eslint/js/issues/488

**Comment by overlookmotel** [comment]
Espree has decided to extend the range of `Program` to cover the entirety of source text. https://github.com/eslint/js/issues/648#issuecomment-2880611833

However, they consider this is breaking change, so it'll be a while.

What does TS-ESLint want to do in the meantime? As I said above:

> Espree sets the end location as the last token, also excluding whitespace and comments. However, `@typescript-eslint/parser` does not match that behavior - end location of `Program` is the end of the file.
>
> I assume that `@typescript-eslint/parser` is aiming to align with Espree, and therefore this difference is unintentional.

Would you like to fix this misalignment with Espree, and consider it a bugfix (i.e. not a breaking change)?

**Comment by overlookmotel** [comment]
ESLint are making the change to `range` of `Program` in ESLint v10. https://eslint.org/blog/2025/10/whats-coming-in-eslint-10.0.0/

I assume that TS-ESLint will want to follow suit. If so, how would you want to time making the change? In next major version after ESLint v10 is released? Or would you consider the current behavior a bug and fix it in a patch release?

One factor that may have bearing on that question is that at present [as noted above](https://github.com/typescript-eslint/typescript-eslint/issues/11026#issue-2968950984) TS-ESLint does not align with Espree's behavior - either the old one, *or* the new one!

Maybe one option is to detect the version of ESLint that TS-ESLint is running in (how?) and align TS-ESLint's behavior with that ESLint version.

**Comment by bradzacher** [comment]
I don't think it's a widely applicable enough thing to be classed as a breaking change.

It's a very specific set of circumstances that cause the range mismatch. There's an additional rarity in that it only actually impacts those that choose to do a mutation on the code based on `program.range[0]` or `program.range[1]`. I very, very, very highly doubt there's really any rules that will break with this change.

So I personally don't mind just landing this in a minor once we bump our supported ESLint version to v10

**Comment by aoto-tech** [comment]
I started looking into this and have a small implementation that aligns `Program.range` with the ESLint v10 / Espree 11 behavior.

The minimal change is quite small: one change in `convert.ts` plus a focused `parse()` regression test, for a 2-file diff.

However, I also tried adding/updating `ast-spec` coverage. Since this change affects the serialized `Program` node for many existing fixtures, regenerating the full ast-spec snapshots turns into a very large mechanical diff: around 331 files changed.

Would you prefer the PR to stay small with just the focused parser regression test, or should it include the broader ast-spec snapshot updates as well?
